crates/red_knot_python_semantic/src/types.rs

```diff
@@ -550,7 +552,7 @@ impl<'db> Type<'db> {
 
         match (self, target) {
             // We should have handled these immediately above.
-            (Type::Dynamic(_), _) | (_, Type::Dynamic(_)) => {
+            (Type::Dynamic(_) | Type::TypeVar(_), _) | (_, Type::Dynamic(_) | Type::TypeVar(_)) => {
```

Comment:
I have implemented all of these `Type` functions with the interpretation that a typevar is not fully static, since it can be specialized to a dynamic type.

Reply:
We are actually going to do a simple syntactic check to determine if a typevar is fully static. The typevar is not fully static iff it has a not-fully-static bound or constraint. This lines up nicely with the definition in the [typing spec](https://typing.python.org/en/latest/spec/concepts.html#fully-static-and-gradual-types):

> We will refer to types that do not contain a gradual form as a sub-part as fully static types.